### PR TITLE
Verify most operations against securitymanager

### LINKED ISSUE #2487: Decouple SecurityManager and Cache and always check with SecurityManager
Currently SecurityManager is only queried if the Cache is used, which elides checks for various commands, e.g. list, key management, extract, mount, ...

The debug commands may be excluded.

### PR DESCRIPTION
**Pull request by enkore**
Fixes #2487 

TODO: Check if some corner case(s) were overlooked. Some functions are now separated and can now be covered with additional tests (the reconanlinzlianoztion thing).

### CODE REVIEW

File: src/borg/cache.py [hunk 1]
```diff
@@ -77,25 +97,27 @@ def assert_location_matches(self, cache):
...
                 raise Cache.RepositoryAccessAborted()
             # adapt on-disk config immediately if the new location was accepted
             logger.debug('security: updating location stored in cache and security dir')
             with open(self.location_file, 'w') as fd:
-                fd.write(cache.repository._location.canonical_path())
-            cache.begin_txn()
-            cache.commit()
+                fd.write(repository_location)
+            if cache_config:
+                cache_config.save()
```
Comment (enkore) [inline src/borg/cache.py]: = Now only the config is updated, not the entire cache transacted (=chunks cache, files cache copied then rewritten then copies deleted).

File: src/borg/cache.py [hunk 2]
```diff
@@ -115,29 +138,150 @@ def assert_no_manifest_replay(self, manifest, key, cache):
...
+            and \
+            cl.path and rl.path and \
+            cl.path.startswith('/~/') and rl.path.startswith('/./') and cl.path[3:] == rl.path[3:]:
+        # everything is same except the expected change in relative path canonicalization,
+        # update previous_location to avoid warning / user query about changed location:
+        return repo_location
+    else:
+        return cache_location
+
+
```
Comment (enkore) [inline src/borg/cache.py]: ^^^

This is all really far more complex than this kind of code ought to be, but given the constraints I don't really see many ways to reduce the intrinsic complexity of that stuff. (So yes, pedantically: the code is complex because the requirements are complex).

File: src/borg/cache.py [hunk 3]
```diff
@@ -240,66 +386,27 @@ def create(self):
...
-        self.config = configparser.ConfigParser(interpolation=None)
-        config_path = os.path.join(self.path, 'config')
-        self.config.read(config_path)
-        self._check_upgrade(config_path)
-        self.id = self.config.get('cache', 'repository')
-        self.manifest_id = unhexlify(self.config.get('cache', 'manifest'))
-        self.timestamp = self.config.get('cache', 'timestamp', fallback=None)
-        self.key_type = self.config.get('cache', 'key_type', fallback=None)
-        self.previous_location = self.config.get('cache', 'previous_location', fallback=None)
+        # TODO config open
```
Comment (ThomasWaldmann) [inline src/borg/cache.py]: ^^^

File: src/borg/cache.py [hunk 4]
```diff
@@ -178,25 +322,27 @@ def __init__(self, repository, key, manifest, path=None, sync=True, do_files=Fal
...
+
         # Warn user before sending data to a never seen before unencrypted repository
         if not os.path.exists(self.path):
-            self.security_manager.assert_access_unknown(warn_if_unencrypted, key)
+            self.security_manager.assert_access_unknown(warn_if_unencrypted, manifest, key)
             self.create()
-        self.open(lock_wait=lock_wait)
+
+        self.open()
+        self.security_manager.assert_secure(manifest, key, cache_config=self.cache_config)
```
Comment (ThomasWaldmann) [inline src/borg/cache.py]: as you moved it outside the try/except, an exception in here will not call self.close().

File: src/borg/cache.py [hunk 5]
```diff
@@ -115,29 +138,152 @@ def assert_no_manifest_replay(self, manifest, key, cache):
...
+        if cache_config and cache_config.key_type is not None and cache_config.key_type != str(key.TYPE):
             raise Cache.EncryptionMethodMismatch()
         if self.known() and not self.key_matches(key):
             raise Cache.EncryptionMethodMismatch()
 
-    def assert_secure(self, manifest, key, cache):
-        self.assert_location_matches(cache)
-        self.assert_key_type(key, cache)
-        self.assert_no_manifest_replay(manifest, key, cache)
+    def assert_secure(self, manifest, key, *, cache_config=None, warn_if_unencrypted=True):
```
Comment (ThomasWaldmann) [inline src/borg/cache.py]: `, *,` ?
so that is for callers giving a cache object there and that is not used, but we rather load the cache_config here?

File: src/borg/cache.py [hunk 6]
```diff
@@ -115,29 +138,152 @@ def assert_no_manifest_replay(self, manifest, key, cache):
...
+        if cache_config:
+            self._assert_secure(manifest, key, cache_config)
+        else:
+            cache_config = CacheConfig(self.repository)
+            if cache_config.exists():
+                with cache_config:
+                    self._assert_secure(manifest, key, cache_config)
+            else:
+                self._assert_secure(manifest, key)
+        logger.debug('security: all checks ok')
```
Comment (ThomasWaldmann) [inline src/borg/cache.py]: that's a rather wide claim. maybe "repository security"?

File: src/borg/cache.py [hunk 7]
```diff
@@ -115,29 +138,152 @@ def assert_no_manifest_replay(self, manifest, key, cache):
...
-            if not yes(msg, false_msg="Aborting.", invalid_msg="Invalid answer, aborting.",
-                       retry=False, env_var_override='BORG_UNKNOWN_UNENCRYPTED_REPO_ACCESS_IS_OK'):
+            allow_access = not warn_if_unencrypted or yes(msg, false_msg="Aborting.",
+                invalid_msg="Invalid answer, aborting.",
+                retry=False, env_var_override='BORG_UNKNOWN_UNENCRYPTED_REPO_ACCESS_IS_OK')
+            if allow_access:
+                if warn_if_unencrypted:
+                    logger.debug('security: saving state for unknown unencrypted repository (explicitly granted)')
+                else:
+                    logger.debug('security: saving state for unknown unencrypted repository')
```
Comment (ThomasWaldmann) [inline src/borg/cache.py]: how about "security: allowing access to unknown unencrypted repository [(explicitly granted)] and remembering decision"

File: src/borg/cache.py [hunk 8]
```diff
@@ -115,29 +138,152 @@ def assert_no_manifest_replay(self, manifest, key, cache):
...
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+    def exists(self):
+        return os.path.exists(self.config_path)
+
+    def create(self):
+        assert not os.path.exists(self.config_path)
```
Comment (ThomasWaldmann) [inline src/borg/cache.py]: assert not self.exists()

Comment (ThomasWaldmann) [inline src/borg/cache.py]: or even:
```
if self.exists():
    raise ...
```

File: src/borg/cache.py [hunk 9]
```diff
@@ -115,29 +138,152 @@ def assert_no_manifest_replay(self, manifest, key, cache):
...
+            config.write(fd)
+
+    def open(self):
+        self.lock = Lock(os.path.join(self.path, 'lock'), exclusive=True, timeout=self.lock_wait,
+                         kill_stale_locks=hostname_is_unique()).acquire()
+        self.load()
+
+    def load(self):
+        self._config = configparser.ConfigParser(interpolation=None)
+        config_path = os.path.join(self.path, 'config')
```
Comment (ThomasWaldmann) [inline src/borg/cache.py]: we already have computed self.config_path.

File: src/borg/cache.py [hunk 10]
```diff
@@ -115,29 +138,152 @@ def assert_no_manifest_replay(self, manifest, key, cache):
...
+            self.previous_location = None
+
+    def save(self, manifest=None, key=None):
+        if manifest:
+            self._config.set('cache', 'manifest', manifest.id_str)
+            self._config.set('cache', 'timestamp', manifest.timestamp)
+        if key:
+            self._config.set('cache', 'key_type', str(key.TYPE))
+        self._config.set('cache', 'previous_location', self.repository._location.canonical_path())
+        with SaveFile(os.path.join(self.path, 'config')) as fd:
```
Comment (ThomasWaldmann) [inline src/borg/cache.py]: self.config_path

File: src/borg/cache.py (hunk 5 above)
Comment (enkore) [inline src/borg/cache.py]: `, *,` ends the positional argument list. Arguments beyond the `*` marker are only accepted as keyword arguments.

**Comment by enkore** [comment]
Squashed; before squashing: 4ca5892
